### PR TITLE
tool to generate constraint approximation databases

### PR DESCRIPTION
**Pull request by v4hn**
This feature was on my todo list for quite some time now.
MoveIt provides [some C++ demo code](https://github.com/ros-planning/moveit/blob/melodic-devel/moveit_planners/ompl/ompl_interface/src/demo_construct_state_database.cpp) to generate constraint approximations.
These can be used for [constraint path planning](https://ros-planning.github.io/moveit_tutorials/doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.html).
The demo is all hard-coded though and users had to copy&paste the code for their own use case.

This pull-request provides a way to [describe constraints on the parameter server](https://github.com/ros-planning/moveit/commit/783cf74a76350692877e6327df247ea4b18f6861) and applies the feature to configure [a generic tool](https://github.com/ros-planning/moveit/commit/3fc0789c6c6da9573cc5134ab2b11e39c7bc5202) to generate the databases.

I'm open to change & move parts of this code if someone has a better proposal for where things should go.

Please look at the individual commits for more details.

### CODE REVIEW

**Comment by v4hn** [comment]
I plan to extend [the constraint manifold tutorial](https://ros-planning.github.io/moveit_tutorials/doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.html) once I got an approval on the current design.

I'm not sure where to document the structure for the `constructConstraints` function best, or whether it needs to be documented explicitly when the manifold-tutorial gives examples already.
Maybe it makes sense to add a link in the function's doxygen documentation?

Thanks for the feedback, I will address it later this week.

File: moveit_core/kinematic_constraints/src/utils.cpp [hunk 1]
```diff
@@ -269,4 +272,230 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
...
+      constraint.position = parseDouble(it->second);
+    }
+    else if (it->first == "tolerance")
+    {
+      constraint.tolerance_below = parseDouble(it->second);
+      constraint.tolerance_above = parseDouble(it->second);
+    }
+    else if (it->first == "tolerances")
+    {
+      if (it->second.getType() != XmlRpc::XmlRpcValue::TypeArray || it->second.size() != 2)
```
[the remark this answers is not in the document]
Comment (v4hn) [reply moveit_core/kinematic_constraints/src/utils.cpp]: Good idea, but I wanted to leave the possibility to possibly check for multiple different structures in the future.

I factorized a `isArray` and an `isStruct` method, but made the logging optional.

File: moveit_core/kinematic_constraints/src/utils.cpp [hunk 2]
```diff
@@ -269,4 +272,230 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
...
+      {
+        ROS_ERROR_STREAM_NAMED(LOGNAME, it->first << " is no struct with x/y/z keys");
+        return false;
+      }
+      constraint.constraint_region.primitive_poses.emplace_back();
+      constraint.constraint_region.primitives.emplace_back();
+      geometry_msgs::Pose& region_pose = constraint.constraint_region.primitive_poses.back();
+      shape_msgs::SolidPrimitive& region_primitive = constraint.constraint_region.primitives.back();
+
+      std::function<double(XmlRpc::XmlRpcValue&)> center = [](XmlRpc::XmlRpcValue& it) {
```
[the remark this answers is not in the document]
Comment (v4hn) [reply moveit_core/kinematic_constraints/src/utils.cpp]: I prefer to keep this as-is for better readability.

File: moveit_core/kinematic_constraints/src/utils.cpp [hunk 3]
```diff
@@ -269,4 +272,230 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
...
+    {
+      ROS_ERROR_STREAM_NAMED(LOGNAME, "orientation constraint contains unknown entity '" << it->first << "'");
+    }
+  }
+  return true;
+}
+
+bool constructConstraint(XmlRpc::XmlRpcValue& params, moveit_msgs::VisibilityConstraint& constraint)
+{
+  ROS_FATAL_NAMED(LOGNAME, "TODO: VisibilityConstraint parsing from parameters is not implemented");
```
[the remark this answers is not in the document]
Comment (v4hn) [reply moveit_core/kinematic_constraints/src/utils.cpp]: To be honest I believe the current implementation of this constraint is far from elegant and I have not seen a single case where this was used in practice..
So I just did not bother before.

I added (and verified) code for this in another commit now.

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  ros::NodeHandle nh("~");
+
+  planning_scene_monitor::PlanningSceneMonitor psm(ROBOT_DESCRIPTION);
+  if (!psm.getRobotModel())
+  {
+    return 1;
+  }
+  if (nh.param<bool>("use_current_scene", true))
+  {
+    ROS_INFO("Requesting current planning scene to generate database");
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: NAMED

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    if (!psm.requestPlanningSceneState())
+    {
+      ROS_FATAL("Abort. The current scene could not be retrieved.");
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: NAMED

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  // - disable edges by default
+  // - local planning in JointModel state space
+  opt.edges_per_sample = nh.param<int>("edges_per_sample", 0);
+  opt.max_edge_length = nh.param<double>("max_edge_length", 0.2);
+  opt.explicit_motions = nh.param<bool>("explicit_motions", true);
+  opt.explicit_points_resolution = nh.param<double>("explicit_points_resolution", 0.05);
+  opt.max_explicit_points = nh.param<int>("max_explicit_points", 200);
+  opt.state_space_parameterization = nh.param<std::string>("state_space_parameterization", "JointModel");
+
+  ROS_INFO_STREAM("Generating Joint Space Constraint Approximation Database for constraint:\n" << constraints);
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: NAMED

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+  const std::string planning_group = nh.param<std::string>("planning_group", "<unspecified planning group>");
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: Is this actually a good default? Maybe return with an error msg would be better here

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+  const std::string output_folder = nh.param<std::string>("output_folder", "constraint_approximations_database");
+  ompl_interface::ConstraintApproximationConstructionResults result =
+      ompl_interface.getConstraintsLibrary().addConstraintApproximation(constraints, planning_group, scene, opt);
+  if (!result.approx)
+  {
+    ROS_FATAL("Failed to generate approximation.");
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: NAMED

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp (hunk 6 above)
Comment (mlautman) [reply moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: Why in the middle of reading parameters? Why not keep all parameters together?

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+      return 1;
+    }
+  }
+
+  XmlRpc::XmlRpcValue constraint_description;
+  moveit_msgs::Constraints constraints;
+  if (!nh.getParam(CONSTRAINT_PARAMETER, constraint_description) ||
+      !kinematic_constraints::constructConstraints(constraint_description, constraints))
+  {
+    ROS_FATAL_STREAM("Could not find valid constraint description in parameter '"
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: NAMED

File: moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  scene->getCurrentStateNonConst().update();
+
+  ompl_interface::OMPLInterface ompl_interface(scene->getRobotModel());
+
+  ompl_interface::ConstraintApproximationConstructionOptions opt;
+
+  // number of states in joint space approximation
+  opt.samples = nh.param<int>("state_cnt", 10000);
+
+  // useful default parameterization
```
Comment (mlautman) [inline moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp]: Seems like making a class and putting all the parameter reading into a separate method would be cleaner

**Comment by v4hn** [comment]
@rhaschke @mlautman I addressed all feedback. Please have another look.

@mlautman Personally, I prefer in-place `nh.param` calls over parameter structs in such small programs, but I switched to a struct here.

File: moveit_core/kinematic_constraints/src/utils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/kinematic_constraints/src/utils.cpp]: For me, your version wasn't readable. It took me a while to figure out, that x,y,z components are all "parsed" twice, once with `center()` and once with `extend()`. Splitting this into two functions disguises the fact that center and extend are both parsed from the very same x,y,z elements, which are ranges actually.

File: moveit_core/kinematic_constraints/src/utils.cpp [hunk 11]
```diff
@@ -273,3 +274,298 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
...
+    {
+      if (!isStruct(it->second, { "x", "y", "z" }, "region"))
+        return false;
+      constraint.constraint_region.primitive_poses.emplace_back();
+      constraint.constraint_region.primitives.emplace_back();
+      geometry_msgs::Pose& region_pose = constraint.constraint_region.primitive_poses.back();
+      shape_msgs::SolidPrimitive& region_primitive = constraint.constraint_region.primitives.back();
+
+      std::function<double(XmlRpc::XmlRpcValue&)> center = [](XmlRpc::XmlRpcValue& it) {
+        return parseDouble(it[0]) + (parseDouble(it[1]) - parseDouble(it[0])) / 2;
```
Comment (rhaschke) [inline moveit_core/kinematic_constraints/src/utils.cpp]: This is hard to read as well. The center of the interval is just `(parseDouble(it[0]) + parseDouble(it[1])) / 2;`

File: moveit_core/kinematic_constraints/src/utils.cpp [hunk 12]
```diff
@@ -273,3 +274,298 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
 
   return goal;
 }
+
+static double parseDouble(XmlRpc::XmlRpcValue& v)
```
Comment (rhaschke) [inline moveit_core/kinematic_constraints/src/utils.cpp]: Please provide some const specifiers to the arguments.

Comment (rhaschke) [reply moveit_core/kinematic_constraints/src/utils.cpp]: Please neglect my comment. Obviously XmlRpc doesn't correctly declare constness of cast operations.

Comment (v4hn) [reply moveit_core/kinematic_constraints/src/utils.cpp]: Yes, the `XmlRpc` API is rather badly designed. This shows in a number of places in the request.
`int` for array indices, no iterators for arrays, missing `const` specifies in API functions, ...

File: moveit_core/kinematic_constraints/src/utils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (v4hn) [reply moveit_core/kinematic_constraints/src/utils.cpp]: I guess it depends on whether you look at it as "parse ros-parameters" or as "fill in a full constraint message". I don't care too much, so I changed it as suggested.

**Comment by rhaschke** [comment]
@v4hn Do you want to squash the cleanup commits into the first 4 commits?

**Comment by v4hn** [comment]
I will clean up the commit history and merge.

**Comment by v4hn** [comment]
merged I'll try to get around to add the tutorial soon.

**Pull request merged**